subprojects/core/src/integTest/groovy/org/gradle/api/invocation/GradleLifecycleBeforeProjectEagerExecutionIntegrationTest.groovy

```diff
@@ -0,0 +1,259 @@
+        output.count("lifecycle.beforeProject for root") == 1
+        output.count("lifecycle.beforeProject for a") == 1
+
+        where:
+        secondEagerExecutionBlock << [
+            "subprojects { $projectMutableStateAccess }",
+            "afterEvaluate { allprojects { $projectMutableStateAccess } }"
+        ]
+    }
+
+    def 'lifecycle.beforeProject eager execution can be triggered from project.#api'() {
+        given:
+        settingsFile """
+            rootProject.name = 'root'
+            gradle.lifecycle.beforeProject {
+                println "lifecycle :\$name"
+                ext.foo = "bar :\$name"
+            }
+            include(":a")
+            include(":b")
+        """
+
+        file("a/build.gradle") << ""
+        file("b/build.gradle") << ""
+
+        buildFile"""
+            $invocation { project ->
+                println "access :\${project.name}"
+                println "\${project.foo}"
+            }
+        """
+
+        when:
+        run "help", "-q"
+
+        then:
+        outputContains expectedOutput
+
+        where:
+        api                | invocation                            | expectedOutput
+        "allprojects"      | "allprojects"                         | "lifecycle :root\naccess :root\nbar :root\naccess :a\nlifecycle :a\nbar :a\naccess :b\nlifecycle :b\nbar :b"
+        "subprojects"      | "subprojects"                         | "lifecycle :root\naccess :a\nlifecycle :a\nbar :a\naccess :b\nlifecycle :b\nbar :b"
+        "project"          | "project(':a')"                       | "lifecycle :root\naccess :a\nlifecycle :a\nbar :a\nlifecycle :b"
+        "findProject"      | "configure(findProject(':a'))"        | "lifecycle :root\naccess :a\nlifecycle :a\nbar :a\nlifecycle :b"
+        "getAllprojects"   | "getAllprojects().forEach"            | "lifecycle :root\naccess :root\nbar :root\naccess :a\nlifecycle :a\nbar :a\naccess :b\nlifecycle :b\nbar :b"
+        "getSubprojects"   | "getSubprojects().forEach"            | "lifecycle :root\naccess :a\nlifecycle :a\nbar :a\naccess :b\nlifecycle :b\nbar :b"
+        "getChildProjects" | "getChildProjects().values().forEach" | "lifecycle :root\naccess :a\nlifecycle :a\nbar :a\naccess :b\nlifecycle :b\nbar :b"
+    }
+
+    def 'lifecycle.beforeProject eager execution can be triggered from gradle.allprojects'() {
+        settingsFile """
+            rootProject.name = 'root'
+            gradle.allprojects {
+                println "allprojects :\$name"
+                println "\${foo}"
+            }
+            gradle.lifecycle.beforeProject {
+                println "lifecycle :\$name"
+                ext.foo = "bar :\$name"
+            }
+            include(":a")
+            include(":b")
+        """
+
+        file("a/build.gradle") << ""
+        file("b/build.gradle") << ""
+
+        when:
+        run "help", "-q"
+
+        then:
+        outputContains """
+allprojects :root
+lifecycle :root
```

Comment:
❓ The mutable access inside the `allprojects{}` call back will trigger the new lifecycle callback for the root project.
Should the same behavior be observed for the `gradle.rootProject{}` callback?

I would argue they should behave consistently. In other words `gradle.rootProject{ <logic> }` should behave the same way `gradle.allprojects { if (it.isRoot) { <logic> } }` behaves.